File: ios/Classes/Options.swift
//
//  Options.swift
//  g_faraday
//
//  Created by gix on 2020/12/29.
//

import Foundation

public struct Options {
    
    let raw: [String: Any]?
    
    init(_ value: [String: Any]?) {
        raw = value
    }
    
    public var animated: Bool {
        return get(key: "_faraday.animated", defaultValue: false)
    }
    
    public var present: Bool {
        return get(key: "_faraday.present", defaultValue: false)
    }
    
    public var isFlutterRoute: Bool {
        return get(key: "_faraday.flutter", defaultValue: false)
    }
    
    public func get<T>(key: String, defaultValue: T) -> T {
        return raw?[key] as? T ?? defaultValue
    }
}

File: lib/src/route/options.dart
import 'package:g_json/g_json.dart';

///
class Options {
  final JSON _raw;

  /// raw options
  Map<String, dynamic> get raw => _raw.mapObject ?? {};

  ///
  Options([Map<String, dynamic>? value]) : _raw = JSON(value ?? {});

  /// enable/disable animation
  factory Options.animated() =>
      Options()..add(key: '_faraday.animated', value: true);

  /// show container with/without present
  factory Options.present() =>
      Options()..add(key: '_faraday.present', value: true);

  /// is flutter route
  factory Options.flutterRoute() =>
      Options()..add(key: '_faraday.flutter', value: true);

  /// add more options
  void add({required String key, required dynamic value}) {
    _raw[key] = value;
  }
}
